### ISSUE TITLE
char.trunc fails with NA

### ISSUE THREAD

**Issue by r2evans** [Rdatatable/data.table#6441]
When `datatable.prettyprint.char` is non-null, `char.trunc` does not correctly handle `NA`s in the data.

```r
> options(datatable.prettyprint.char = NULL)
> data.table(a = c("A", NA))
        a
   <char>
1:      A
2:   <NA>

> options(datatable.prettyprint.char = 10)
> data.table(a = c("A", NA))
Error in if (!any(idx)) return(x) : missing value where TRUE/FALSE needed
```

The issue is clear:

```r
 > debug(data.table:::char.trunc)
 > data.table(a = c("A", NA))
 debugging in: char.trunc(x)
 debug: {
     trunc.char = max(0L, suppressWarnings(as.integer(trunc.char[1L])),
         na.rm = TRUE)
     if (!is.character(x) || trunc.char <= 0L)
         return(x)
     nchar_width = nchar(x, "width")
     nchar_chars = nchar(x, "char")
     is_full_width = nchar_width > nchar_chars
     idx = pmin(nchar_width, nchar_chars) > trunc.char
     if (!any(idx))
         return(x)
     x[idx] = paste0(strtrim(x[idx], trunc.char * fifelse(is_full_width[idx],
         2L, 1L)), "...")
     x
 }
 Browse[2]> debug: trunc.char = max(0L, suppressWarnings(as.integer(trunc.char[1L])),
     na.rm = TRUE)
 Browse[2]> debug: if (!is.character(x) || trunc.char <= 0L) return(x)
 Browse[2]> debug: nchar_width = nchar(x, "width")
 Browse[2]> debug: nchar_chars = nchar(x, "char")
 Browse[2]> debug: is_full_width = nchar_width > nchar_chars
 Browse[2]> debug: idx = pmin(nchar_width, nchar_chars) > trunc.char
 Browse[2]> debug: if (!any(idx)) return(x)
 Browse[2]> idx
 [1] FALSE    NA
```

This can be easily fixed with either `if (!any(na.omit(idx)))` or `idx[is.na(idx)] <- FALSE`.

<details><summary> sessionInfo </summary>

```
> sessionInfo()
 R version 4.3.2 (2023-10-31)
 Platform: x86_64-pc-linux-gnu (64-bit)
 Running under: Ubuntu 24.04.1 LTS

 Matrix products: default
 BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
 LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0

 locale:
  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C
 [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C

 time zone: America/New_York
 tzcode source: system (glibc)

 attached base packages:
 [1] stats     graphics  grDevices datasets  utils     methods   base

 other attached packages:
 [1] data.table_1.16.0

 loaded via a namespace (and not attached):
 [1] compiler_4.3.2 tools_4.3.2    renv_1.0.7
```

</details>

It worked as expected in 1.15.2 at least:

```
> library(data.table)
> options(datatable.prettyprint.char = NULL)
> data.table(a = c("A", NA))
        a
   <char>
1:      A
2:   <NA>
> options(datatable.prettyprint.char = 36)
> data.table(a = c("A", NA))
        a
   <char>
1:      A
2:   <NA>
> sessionInfo()
R version 4.3.3 (2024-02-29)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 24.04.1 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8        LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C           LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

time zone: America/New_York
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] data.table_1.15.2 r2_0.11.0        

loaded via a namespace (and not attached):
 [1] compiler_4.3.3    clipr_0.8.0       fastmap_1.2.0     cli_3.6.2         tools_4.3.3       htmltools_0.5.8.1 rmarkdown_2.26    knitr_1.45       
 [9] xfun_0.42         digest_0.6.34     rlang_1.1.3       evaluate_0.23    
```

**Comment by joshhwuu** [comment]
Oversight by me in #6048, probably won't make 1.16.0(?) but I'll file a PR to fix, thanks for flagging + solution!

**Comment by r2evans** [comment]
> thanks a lot for filing this. I was losing my mind!

I admit it took me a bit to differentiate the error in my code processing the frame from the printing of the result.

Now that it's fixed and released as 1.16.2 (mirrors are still catching up), I'm happily testing 1.16.2 again :fireworks:
